### ISSUE TITLE
minIOURandomCrop transorm is possibly buggy

### ISSUE THREAD

**Issue by korabelnikov** [open-mmlab/mmdetection#1582]
https://github.com/open-mmlab/mmdetection/blob/d88ad8fc32a6cce46c72619a16b85226360a7ba2/mmdet/datasets/pipelines/transforms.py#L584

1) docstring is wrong.
2) Seems like such transformation may cause errors because may produce patch of various aspect ratios that will be resized to one on standard formats (VOC/COCO). 
Simply: vertical rectangle may become horizontal and vice-versa. Generally it's wrong and it's better to crop patch of same aspect ratio as full image.

I can prepare PR if someone interested

**Comment by hellock** [comment]
@korabelnikov Thanks for the report and feel free to prepare a PR to fix it! @yhcao6 Please confirm this bug.

**Comment by yhcao6** [comment]
Thanks for reporting the bug.

1. Docstring is wrong. 
It is indeed wrong. There is no `crop_size` in args, it should be `min_crop_size`.

2. `MinIoURandomCrop` will produce patch of various aspect ratios. 
This transform is designed for [SSD](https://github.com/open-mmlab/mmdetection/blob/master/configs/ssd300_coco.py) and we just follow the official implementation.

We are appreciate if you can open a pr to fix it.

**Comment by hhaAndroid** [comment]
If the aspect ratio changes after the transformation, I think it is indeed unreasonable. Because the flag is not changed during group sampling

**Comment by hhaAndroid** [comment]
I think there is a problem with this code, especially:

`  if len(overlaps) > 0 and overlaps.min() < min_iou:  
                    continue`

If there is a small bbox, then this condition will always be satisfied, this way of writing is wrong

The correct one should be:
`     if overlap.max() < min_iou or overlap.min() > max_iou:
              continue`

**Comment by ChenCVer** [comment]
> I think there is a problem with this code, especially:
> 
> ` if len(overlaps) > 0 and overlaps.min() < min_iou: continue`
> 
> If there is a small bbox, then this condition will always be satisfied, this way of writing is wrong
> 
> The correct one should be:
> ` if overlap.max() < min_iou or overlap.min() > max_iou: continue`

right!

**Comment by taofuyu** [comment]
> I think there is a problem with this code, especially:
> 
> ` if len(overlaps) > 0 and overlaps.min() < min_iou: continue`
> 
> If there is a small bbox, then this condition will always be satisfied, this way of writing is wrong
> 
> The correct one should be:
> ` if overlap.max() < min_iou or overlap.min() > max_iou: continue`

But when len(overlaps)==0, overlaps.max() or overlaps.min() will raise ValueError. So the correct one should be:
if len(overlaps)==0 or overlaps.max() < min_iou or overlaps.min() > max_iou: continue

**Comment by hhaAndroid** [comment]
> > I think there is a problem with this code, especially:
> > ` if len(overlaps) > 0 and overlaps.min() < min_iou: continue`
> > If there is a small bbox, then this condition will always be satisfied, this way of writing is wrong
> > The correct one should be:
> > ` if overlap.max() < min_iou or overlap.min() > max_iou: continue`
> 
> But when len(overlaps)==0, overlaps.max() or overlaps.min() will raise ValueError. So the correct one should be:
> if len(overlaps)==0 or overlaps.max() < min_iou or overlaps.min() > max_iou: continue

OK, Thank you. I will check it.

**Comment by ethanyhzhang** [comment]
for the code belows, why calculate the iou between the patch and the bbox using 'iou' mode?  as patch could be much larger than the bboxes.
```
overlaps = bbox_overlaps( patch.reshape(-1, 4), boxes.reshape(-1, 4)).reshape(-1)
```
would it be better like this:
overlaps = bbox_overlaps( boxes.reshape(-1, 4), patch.reshape(-1, 4),  mode='iof').reshape(-1)

**Comment by RangiLyu** [comment]
We did some experiments and found that when changing `overlaps.min()` to `overlaps.max()` will lead to performance degradation(drop about 0.3~0.5 mAP on SSD and 0.2 on YOLOv3). 
Also, we found that the IOU crop operation in the augmentation pipeline is not really where it works. The key is the expand operation before random crop. Combining expand and random crop together is similar to multi-scale training, whether to crop the object with a specific IOU is not so important.
